text/0057-checksummed-addresses.md

```diff
@@ -0,0 +1,167 @@
+# Title
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0057](https://github.com/casperlabs/ceps/pull/0057)
+
+This CEP introduces an opt-in checksum scheme to avoid copy errors.
+The encoding generalizes [EIP-55][1]. We provide a scheme for
+hex-encoded values. The approach encodes a checksum in the
+capitalization of hexadecimal digits. While the new format protects
+account hashes, it also protects all hex-encoded values.
+
+[1]: https://eips.ethereum.org/EIPS/eip-55
+
+## Motivation
+
+[motivation]: #motivation
+
+Copy errors when entering account addresses may result in the loss of
+millions. A checksum scheme avoids this by including redundant
+information.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The scheme encodes a checksum in the letter capitalization of hex-encodings.
+It extends [EIP-55][1] to work with arbitrary length arrays.
+
+The encoding uses the hash of the input to determine capitalization.
+The hash is broken into 4-bit chunks called *nibbles*. A nibble
+corresponds to a hexadecimal digit. The procedure also breaks the
+input into nibbles. Each input nibble uses the hash nibble with the
+same index for capitalization. If the corresponding hash nibble is
+less than or equal to 7, the hex character is capitalized. Otherwise,
+it is lower case.
+
+Decoding verifies the correct capitalization of each letter. The
+decoder skips verification if all letters are lower case or upper
+case.
+
+[1]: https://eips.ethereum.org/EIPS/eip-55
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The design leverages two special `encode` and `decode` function which
+work on hexadecimal strings.
+
+### Converting Bytes to Nibbles
+
+The encoding procedure relies a helper function which converts bytes
+to nibbles. As mentioned in the [guide-level explanation](#guide-level-explanation), nibbles are 4-bit chunks. They
+correspond to hexadecimal digits.
+
+```rust
+fn bytes_to_nibbles(bytes: &[u8]) -> Vec<u8> {
+  let output_nibble_count = bytes.len() * 2;
+  let mut output_nibbles = Vec::with_capacity(output_nibble_count);
+  for i in 0..output_nibble_count {
+    let byte_value = bytes[i / 2];
+    let nibble = byte_value
+            .checked_shr((((i + 1) % 2) * 4) as u32)
+            .unwrap_or(0)
+            & 0xf;
+    output_nibbles.push(nibble);
+  }
+  output_nibbles
```

Comment:
Maybe simpler:
```rust
bytes.iter().flat_map(|byte| {
    [(byte & 0xf0) >> 4, byte & 0x0f]
}).collect()
```

Reply:
So I couldn't get what you wrote above to work, but this works... albeit with a lot of heap allocations:

```rust
fn bytes_to_nibbles(input: &(impl AsRef<[u8]> + ?Sized)) -> Vec<u8> {
    let bytes: &[u8] = input.as_ref();
    bytes
        .iter()
        .flat_map(|byte| vec![(byte & 0xf0) >> 4, byte & 0x0f])
        .collect()
}
```

This isn't *quite* as short, but it's in the same spirit:

```rust
fn bytes_to_nibbles(input: &(impl AsRef<[u8]> + ?Sized)) -> Vec<u8> {
    let bytes: &[u8] = input.as_ref();
    let mut output_nibbles = Vec::with_capacity(bytes.len() * 2);
    for b in bytes {
        output_nibbles.push(b >> 4);
        output_nibbles.push(b & 0x0f);
    }
    output_nibbles
}
```